H0: Tips_M = Tips_F
Ha = Tips_M =/= Tips_F
p-value = probability of observing the result assuming H_0 is true
Significance level (alpha) = Threshold for rejecting H_0

a small p value (<0.05) means a real effect is likely present so you reject H0
a large p value means that we cannot conclude a differenc eexists so we do not reject h0
p value does not prove anything, only provides evidence against H_0

In [2]:
import pandas as pd 

tips = pd.read_csv("tips.csv")

tips.head(10)


ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject

In [ ]:
tips["time"].value_counts()

time
Dinner    176
Lunch      68
Name: count, dtype: int64

In [ ]:
tips.groupby("sex")["tip"].describe()

count      mean       std  min  25%   50%   75%   max
sex                                                          
Female   87.0  2.833448  1.159495  1.0  2.0  2.75  3.50   6.5
Male    157.0  3.089618  1.489102  1.0  2.0  3.00  3.76  10.0

H_0 = There is no difference
H_A =  THere is a difference i.e Tips_M =/= Tips_F

In [ ]:
from scipy.stats import ttest_ind

#split data by gender

male_tips = tips[tips["sex"] == "Male"]["tip"]
female_tips = tips[tips["sex"] == "Female"]["tip"]

print(male_tips)

1      1.66
2      3.50
3      3.31
5      4.71
6      2.00
       ... 
236    1.00
237    1.17
239    5.92
241    2.00
242    1.75
Name: tip, Length: 157, dtype: float64


In [ ]:
#perform t-test (welch's t test)

t_stat, p_value = ttest_ind(male_tips, female_tips, equal_var=False)

print(t_stat)
print(p_value)

#on 5% significance level

if p_value < 0.05:
    print("Reject H_0, an effect/diff is likely present, suggesting difference in tipping behaviors")
else: 
    print("Do not reject H_0, no strong evidence of difference in tipping behavior")



1.489536377092501
0.13780683808650296
Do not reject H_0, no strong evidence of difference in tipping behavior


A p-value of 0.13 means that if the null hypothesis is true, there is a 13% probability of obtaining a result as extreme as (or more extreme than) the observed data just by random chance.



Logistic Regression

Predicting if customer is smoker based on total bill or tips

Logistic Regression uses a sigmoid function

In [ ]:
import numpy as np
import matplotlib as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

tips["smoker"] = tips["smoker"].map({"Yes": 1, "No":0})
tips["sex"] = tips["sex"].map({"Male": 1, "Female":0})

tips.head(10)

total_bill   tip  sex  smoker  day    time  size
0       16.99  1.01    0       0  Sun  Dinner     2
1       10.34  1.66    1       0  Sun  Dinner     3
2       21.01  3.50    1       0  Sun  Dinner     3
3       23.68  3.31    1       0  Sun  Dinner     2
4       24.59  3.61    0       0  Sun  Dinner     4
5       25.29  4.71    1       0  Sun  Dinner     4
6        8.77  2.00    1       0  Sun  Dinner     2
7       26.88  3.12    1       0  Sun  Dinner     4
8       15.04  1.96    1       0  Sun  Dinner     2
9       14.78  3.23    1       0  Sun  Dinner     2

In [ ]:
tips.groupby("sex")["smoker"].value_counts()

Series([], Name: count, dtype: int64)

In [ ]:
X = tips[["total_bill", "tip", "sex"]]
Y = tips["smoker"]
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

print(f"Training samples: {len(X_train)}, Testing samples: {len(X_test)}")

Training samples: 195, Testing samples: 49


In [ ]:
lg_model = LogisticRegression()
lg_model.fit(X_train, Y_train)
y_pred = lg_model.predict(X_test)

In [ ]:
print("Predicted: ", y_pred[:10])
print("Actual: ", Y_test[:10])

Predicted:  [0 0 1 0 0 0 1 0 0 0]
Actual:  24     0
6      0
153    0
211    1
198    1
176    1
192    1
124    0
9      0
101    1
Name: smoker, dtype: int64


In [ ]:
accuracy = accuracy_score(Y_test, y_pred)
print(accuracy)

class_report = classification_report(Y_test, y_pred)
print(class_report)

0.5918367346938775
              precision    recall  f1-score   support

           0       0.63      0.87      0.73        31
           1       0.33      0.11      0.17        18

    accuracy                           0.59        49
   macro avg       0.48      0.49      0.45        49
weighted avg       0.52      0.59      0.52        49



Accuracy = correct predictions / total predictions

Precision: when the model predicts non smokers, it is 64% right, when it predicts as smokers, its 40% right
= TP/(TP+FP) (HOW MANY OF THE PREDICTED POSITIVE CASES WERE ACTUALLY CORRECT)
Recall: out of all actual non smokers, the model correctly identifies 90% of them, out of all smokers, model correctly identitifeies 11% of them
=TP/(TP+FN) (MEASURES HOW MANY ACTUAL POSIITVE CASES WERE CORRECTLY IDENTIFIED)

f1 score is a balance beetween precisipn and recall